File: Leetcode/53. Maximum Subarray.py


# Link - https://leetcode.com/problems/maximum-subarray/

# Solution 1 - Kadane
    # Space: O(1)
    # Time: O(n)

class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        
        sum_so_far = 0
        start = end = 0

        ans = float('-inf')

        while (end < len(nums)):
            sum_so_far += nums[end]

            ans = max(sum_so_far, ans)

            if sum_so_far < 0:
                start += 1
                start = end + 1
                sum_so_far = 0
            
            end += 1
        
        return ans

# Solution 2 - DP
    # Space: O(1)
    # Time: O(n)

class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        
        dp = nums[0]
        M = dp
        
        for i in range(1, len(nums)):
            dp = nums[i] + (dp if dp > 0 else 0)
            M = max(M, dp)
        
        return M

File: Leetcode/53_Maximum_Subarray.cpp
/*
    https://leetcode.com/problems/maximum-subarray/
    Time: O(n)
    Space: O(1)
*/

class Solution
{
    /*
    Kadane's algorithm: don't keep negative subarrays
*/
public:
    int maxSubArray(vector<int> &nums)
    {

        int res = -1e5, sum = 0, i = 0, j = 0;

        while (j < nums.size())
        {

            sum += nums[j];
            res = max(res, sum);

            j++;

            if (sum <= 0)
            {
                i = j;
                sum = 0;
            }
        }

        return res;
    }
};

class Solution
{
    /*
        dp[i] = max of subarray that ends at i
    */
public:
    int maxSubArray(vector<int> &nums)
    {
        int res = nums[0], dp = nums[0];

        for (int i = 1; i < nums.size(); ++i)
        {
            dp = nums[i] + max(dp, 0);
            res = max(res, dp);
        }

        return res;
    }
};